### DISCUSSION TITLE
Kronecker products not DPP

### DISCUSSION THREAD

**Discussion by JAT38** [cvxpy/cvxpy#2483] [Q&A]
Hello,

I am solving a biconvex problem by alternating between two convex problems. I'd like to use parameters to speed up the solution of both.

One of the problems is not DPP because I take some Kronecker products (cvxpy.kron) between parameter matrices and constant numpy matrices. It seems to me the result of this should be parameter affine. But I see in the code that the Kronecker product is never DPP.

One workaround is to use for loops with a bunch of hstacks and vstacks, but this is very slow.

Any reason Kronecker products aren't DPP? I'm basically using them for what repeat does in Numpy. Any obvious tools I'm missing here?

Josh

**Comment by phschiele** [comment]
@JAT38 This is a great question. You're right, there is no mathematical reason stopping kron to be DPP compliant. 
The reason it is not at the moment is that the current default C++ backend is hard to work with and extend. Adding DPP kron to the newer Python backends wouldn't be hard, but it would leave us with different capabilities for the backends, and we have no logic to support this divergence in capabilities at the moment. So solutions would be:
1. Someone takes up the challenge to add DPP kron to the C++ backend
2. We wait until the C++ backend can be retired (this could be a fair bit in the future)
3. We build logic to handle diverging capabilities of the backends

I think 3. could also be required for the N-dimension support that @Transurgeon and @SteveDiamond are working on.

**Comment by JAT38** [comment]
Thank you for your thoughts. This is not a showstopper for me. It's nice to see CVXPy getting better and better.

**Reply by JAT38** [reply]
Just following up to say I came back to this problem a couple days ago. I should have realized before that it's all doable with vec, outer products, and reshape. All of these are DPP and not noticeably less efficient than kron.
